### PR TITLE
Implement delete functionality for follow feature

### PR DESCRIPTION
**Pull request by lukasberglund**
The idea here is to make it possible for the user to unfollow a business.

### CODE REVIEW

File: capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java [hunk 1]
```diff
@@ -101,7 +114,14 @@ private boolean doesBusinessExist(String businessId) {
     }
   }
 
-  private boolean followAlreadyStored(String userId, String businessId) {
+  private boolean followExistsInDatastore(String userId, String businessId) {
+    int numberOfMatches =
+        prepareFollowQuery(userId, businessId).countEntities(withLimit(ENTITY_COUNTING_LIMIT));
+
+    return numberOfMatches > 0;
+  }
```
Comment (eash130) [inline capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java:122]: Just to confirm, there won't be a case where there are multiple pairs of the same userId + businessId correct?

File: capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java [hunk 2]
```diff
@@ -112,9 +132,43 @@ private boolean followAlreadyStored(String userId, String businessId) {
...
+      response.sendError(
+          HttpServletResponse.SC_BAD_REQUEST,
+          "Please specify the ID of the business you would like to unfollow.");
+      return;
+    }
+
+    String userId = currentUser.getUserId();
+    if (!followExistsInDatastore(userId, businessId)) {
+      response.sendError(
+          HttpServletResponse.SC_NOT_FOUND,
```
Comment (eash130) [inline capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java:159]: status code SC_FORBIDDEN seems to be a better fit here.

File: capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java (hunk 1 above)
Comment (lukasberglund) [reply capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java:122]: Not if this works.

File: capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java (hunk 2 above)
Comment (lukasberglund) [reply capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java:159]: I'm not sure I agree... It's not that the user is forbidden from unfollowing, but more that the user simply can't unfollow a business that she isn't following.

File: capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java [hunk 3]
```diff
@@ -49,9 +52,17 @@
...
   private final int ENTITY_COUNTING_LIMIT = 1;
 
+  private int entityListLimit = 100;
   private UserService userService = UserServiceFactory.getUserService();
   private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
 
+  public FollowServlet() {}
+
+  /** Construtor provided for testing */
+  public FollowServlet(int entityListLimit) {
```
Comment (alygerharter) [inline capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java:62]: I really would prefer not to have a test-only constructor.

File: capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java (hunk 2 above)
Comment (alygerharter) [reply capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java:159]: NOT_FOUND is appropriate here.

File: capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java [hunk 4]
```diff
@@ -112,9 +132,43 @@ private boolean followAlreadyStored(String userId, String businessId) {
...
+    }
+
+    while (followExistsInDatastore(userId, businessId)) {
+      List<Entity> followsToDelete =
+          prepareFollowQuery(userId, businessId)
+              .asList(FetchOptions.Builder.withLimit(entityListLimit));
+      for (Entity follow : followsToDelete) {
+        datastore.delete(follow.getKey());
+      }
+    }
```
Comment (alygerharter) [inline capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java:170]: Why the list and loop? Shouldn't there only be one entry in the datastore to follow?

File: capstone/src/test/java/com/google/sps/servlets/follow/FollowServletTest.java [hunk 5]
```diff
@@ -205,4 +216,115 @@ public void testUserCantFollowSelf() throws IOException {
...
+
+    assertResponseWithArbitraryTextRaised(HttpServletResponse.SC_BAD_REQUEST, response);
+  }
+
+  /**
+   * If there are multiple follows with the same property (which shouldn't happen,) the servlet
+   * should delete all of them.
+   */
+  @Test
+  public void testDeletesAllFollows() throws IOException {
```
Comment (alygerharter) [inline capstone/src/test/java/com/google/sps/servlets/follow/FollowServletTest.java:299]: No need to write a test for something that can't happen. You've already established with the doPost tests that this won't happen.

File: capstone/WEB-INF/datastore-indexes.xml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="utf-8"?>
+<datastore-indexes
+  autoGenerate="true">
+
+    <datastore-index kind="Comment" ancestor="false" source="manual">
+        <property name="businessId" direction="asc"/>
+        <property name="parentId" direction="asc"/>
+        <property name="timestamp" direction="desc"/>
+    </datastore-index>
+</datastore-indexes>
```
Comment (winniewan) [inline capstone/WEB-INF/datastore-indexes.xml:10]: Add a new line

File: capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java [hunk 7]
```diff
@@ -49,8 +52,9 @@
 public class FollowServlet extends HttpServlet {
   private final int ENTITY_COUNTING_LIMIT = 1;
 
+  private int entityListLimit = 100;
   private UserService userService = UserServiceFactory.getUserService();
-  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();re
```
Comment (winniewan) [inline capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java:57]: typo: re

Comment (lukasberglund) [reply capstone/src/main/java/com/google/sps/servlets/follow/FollowServlet.java:57]: I wonder why that didn't lead to an error...

**Pull request merged**